_posts/en/newsletters/2020-02-05-newsletter.md

```diff
@@ -0,0 +1,278 @@
+public key may leak the secret key through an (invalid) signature."
+
+1/27 18:36 #secp256k1
+[6:36:02 pm] <gmaxwell> but it's not really a problem in ed25519 because the only thing it does with the public key is stuffs it into the e hash, and presumably it stuffs the same data into the nonce hash?  (if it doesn't they're pants on head stupid and should be shot)
+[...]
+[6:37:49 pm] <gmaxwell> okay the ed25519 people should be shot, it's an input to e and not an input to their nonce function
+[6:38:02 pm] <gmaxwell> (I just checked)
+
+1/28 01:49 https://moderncrypto.org/mail-archive/curves/2020/001012.html
+-->{% endcomment %}
+
+- **Safety concerns related to precomputed public keys used with schnorr signatures:** [libsecp256k1 PR
+  #558][libsecp256k1 #558] proposes adding [BIP340][]-compatible
+  [schnorr signature][topic schnorr signatures] creation and
+  verification to the libsecp256k1 library used by Bitcoin Core and
+  several other Bitcoin programs.  Since BIP340 requires signatures
+  to commit to the public key that will be used to validate the signature,
+  the current proposed signature function uses the private key to derive
+  the necessary public key.  Gregory Maxwell [noted][gmaxwell pubkey]
+  that programs generating signatures will usually already know the
+  appropriate public key, so the function can save CPU time by accepting
+  the public key as a parameter.
+
+    Jonas Nick [replied][nick nonce] that the approach was reasonable,
+    but doing it safely would require including the public key in the
+    data used to create a deterministic nonce.  Otherwise, if an attacker
+    could obtain two signatures created by the same private key for two
+    different public keys, and all of the other data remained the same,
+    you would unknowingly be reusing a nonce and the attacker could
+    derive your private key and steal your bitcoins.  Discussion about
+    addressing problem would continue in a [separate issue][nonce
+    issue].
+
+    Additionally, as it became clear that implementations that accept
+    public keys without verifying them could end up producing reused
+    nonces, Gregory Maxwell [posted][curves post] to a mailing list
+    dedicated to [ed25519][] implementations (which also use a variation
+    of schnorr signatures) about the risk.  Ed25519 co-author Daniel
+    Bernstein replied that "the standard defense against faults is for
+    the signer to verify each signature."  This would detect that an
+    invalid public key had been provided, and some wallets such as
+    Bitcoin Core certainly do perform this check for any signatures they
+    generate even for the currently-used ECDSA signature algorithm.
+    However, there remains a risk that inexperienced developers will not
+    perform this step, and so it seems preferable to use Jonas Nick's
+    suggestion (relayed by Maxwell) of including the public key in the
+    data used to produce a deterministic nonce.
```

Comment:
Just now on #secp256k1 irc:
```text
08:05 <gmaxwell> https://bitcoinops.org/en/newsletters/2020/02/05/ is kinda odd
08:06 <gmaxwell> anyone ever find any implementation that verifies after in open source code in the wild?
08:06 <gmaxwell> I think it's strange to say " However, there remains a risk that inexperienced developers will not perform this step" regarding a pretty computationally expensive step that I don't ever recall seeing any implementation take.
08:08 <sipa> gmaxwell: agree
08:10 <sipa> harding: ping ^... perhaps it's better to say something like "However, the computational overhead of this approach may not be acceptable to many applications" or so
```

Reply:
```text
08:17 <gmaxwell> (I think probably _we_ should do this in libsecp, optionally annd have it controllable with a context flag...because our signing is usually used in cases where performance doesn't matter)
```